optional.py

```diff
@@ -31,6 +33,21 @@ def get(self):
 
         return self._thing
 
+    def map(self, func):
+        if self.is_empty():
```

Comment:
These `is_empty` checks made me create a new branch that has Optional as an ABC with two implementations, _Something and _Nothing.

Reply:
I looked over your branch and it looks good.  Dropped the check lock on get but thats a debatable feature.  I liked the idea of failing early with a not checked error.